core/src/main/java/com/google/zxing/qrcode/encoder/MinimalEncoder.java

```diff
@@ -0,0 +1,778 @@
+                                                                       * (2_BYTE_ISO-8859-2) -- BYTE(Y) (80) --> (3_BYTE_UTF-16BE)
+ * (initial) -- BYTE(X) (40) --> (1_BYTE_UTF-8) -- BYTE(X) (56) --> (2_BYTE_UTF-8)
+ * (initial) -- BYTE(X) (40) --> (1_BYTE_UTF-16BE) -- BYTE(X) (56) --> (2_BYTE_UTF-16BE)
+ * 
+ * Situation after adding edges to vertices at position 3
+ * (initial) -- BYTE(X) (32) --> (1_BYTE_ISO-8859-2) -- BYTE(X) (40) --> (2_BYTE_ISO-8859-2) -- BYTE(Y) (72) --> (3_BYTE_ISO-8859-3)
+                                                                                                               * (3_BYTE_ISO-8859-3) -- BYTE(Y) (80) --> (4_BYTE_ISO-8859-3)
+                                                                                                               * (3_BYTE_ISO-8859-3) -- BYTE(Y) (112) --> (4_BYTE_UTF-8)
+                                                                                                               * (3_BYTE_ISO-8859-3) -- BYTE(Y) (112) --> (4_BYTE_UTF-16BE)
+ * (initial) -- BYTE(X) (40) --> (1_BYTE_UTF-8) -- BYTE(X) (56) --> (2_BYTE_UTF-8) -- BYTE(Y) (72) --> (3_BYTE_UTF-8)
+ * (initial) -- BYTE(X) (40) --> (1_BYTE_UTF-16BE) -- BYTE(X) (56) --> (2_BYTE_UTF-16BE) -- BYTE(Y) (72) --> (3_BYTE_UTF-16BE)
+ * 
+ * Situation after adding edges to vertices at position 4
+ * (initial) -- BYTE(X) (32) --> (1_BYTE_ISO-8859-2) -- BYTE(X) (40) --> (2_BYTE_ISO-8859-2) -- BYTE(Y) (72) --> (3_BYTE_ISO-8859-3) -- BYTE(Y) (80) --> (4_BYTE_ISO-8859-3)
+                                                                                                               * (3_BYTE_UTF-8) -- BYTE(Y) (88) --> (4_BYTE_UTF-8)
+                                                                                                               * (3_BYTE_UTF-16BE) -- BYTE(Y) (88) --> (4_BYTE_UTF-16BE)
+ * (initial) -- BYTE(X) (40) --> (1_BYTE_UTF-8) -- BYTE(X) (56) --> (2_BYTE_UTF-8) -- BYTE(Y) (72) --> (3_BYTE_UTF-8)
+ * (initial) -- BYTE(X) (40) --> (1_BYTE_UTF-16BE) -- BYTE(X) (56) --> (2_BYTE_UTF-16BE) -- BYTE(Y) (72) --> (3_BYTE_UTF-16BE)
+ *
+ * Encoding as ECI(ISO-8859-2),BYTE(XX),ECI(ISO-8859-3),BYTE(YY) has the smallest size of 80 and is hence chosen. The encodation ECI(UTF-8),BYTE(XXYY) is longer with a size of 88.
+ */
+    int inputLength = stringToEncode.length();
+//Array that represents vertices. There is a vertex for every character, encoding and mode. The vertex contains a list of
+//all edges that lead to it that have the same encoding and mode.
+//The lists are created lazily
+    Vector<ResultNode>[][][] vertices = new Vector[inputLength + 1][encoders.length][4];
+    addEdges(version,vertices,0,null);
+    if (DEBUG) {
+        System.err.println("DEBUG Initial situation");
+        printEdges(vertices);
+    }
+    for (int i = 1; i <= inputLength; i++) {
+      for (int j = 0; j < encoders.length; j++) {
+        for (int k = 0; k < 4; k++) {
```

Comment:
Same, what's 4 here

Reply:
There are only 4 modes (KANJI, ALPHANUMERIC, NUMERIC and BYTE) that are relevant in the algorithm. There is a function getCompactedOrdinal(Mode mode) that "compacts" the ordinals of these to values between 0 and 3.